ctapipe/reco/HillasReconstructor.py

```diff
@@ -218,30 +220,58 @@ def initialize_hillas_planes(
-            p2_x = moments.x + 0.1 * u.m * np.cos(moments.psi)
-            p2_y = moments.y + 0.1 * u.m * np.sin(moments.psi)
-            focal_length = subarray.tel[tel_id].optics.equivalent_focal_length
 
             pointing = SkyCoord(
                 alt=telescopes_pointings[tel_id].alt,
                 az=telescopes_pointings[tel_id].az,
                 frame=horizon_frame,
             )
 
-            camera_frame = CameraFrame(
-                focal_length=focal_length, telescope_pointing=pointing
-            )
+            focal_length = subarray.tel[tel_id].optics.equivalent_focal_length
 
-            cog_coord = SkyCoord(x=moments.x, y=moments.y, frame=camera_frame,)
-            cog_coord = cog_coord.transform_to(horizon_frame)
+            if moments.x.unit == u.Unit("m"):  # Image parameters are in CameraFrame
+
+                # we just need any point on the main shower axis a bit away from the cog
+                p2_x = moments.x + 0.1 * u.m * np.cos(moments.psi)
+                p2_y = moments.y + 0.1 * u.m * np.sin(moments.psi)
+
+                camera_frame = CameraFrame(
+                    focal_length=focal_length, telescope_pointing=pointing
+                )
+
+                cog_coord = SkyCoord(x=moments.x, y=moments.y, frame=camera_frame)
+                p2_coord = SkyCoord(x=p2_x, y=p2_y, frame=camera_frame)
+
+            else:  # Image parameters are already in TelescopeFrame
+
+                self.mode = "TelescopeFrame"
 
-            p2_coord = SkyCoord(x=p2_x, y=p2_y, frame=camera_frame)
+                # we just need any point on the main shower axis a bit away from the cog
+                p2_delta_alt = moments.y + 0.1 * u.deg * np.sin(moments.psi)
+                p2_delta_az = moments.x + 0.1 * u.deg * np.cos(moments.psi)
+
+                telescope_frame = TelescopeFrame(telescope_pointing=pointing)
+
+                cog_coord = SkyCoord(
+                    fov_lon=moments.x, fov_lat=moments.y, frame=telescope_frame
+                )
+                p2_coord = SkyCoord(
+                    fov_lon=p2_delta_az, fov_lat=p2_delta_alt, frame=telescope_frame
+                )
+
+            # Coordinates in the sky
+            cog_coord = cog_coord.transform_to(horizon_frame)
             p2_coord = p2_coord.transform_to(horizon_frame)
 
-            # re-project from sky to a "fake"-parallel-pointing telescope
-            # then recalculate the psi angle
+            # DIVERGENT MODE
             if self.divergent_mode:
+
+                # re-project from sky to a "fake"-parallel-pointing telescope
+                # then recalculate the psi angle
+                # In this case there is a posterior transformation to a parallel
+                # so the origin of the coordinates shouldn't matter
+                # WARNING: possibly missing unit-test!
+
                 camera_frame_parallel = CameraFrame(
```

Comment:
I do wonder if this re-computing of the psi angle is still needed, or should be done in the TelescopeFrame since that is now where the HillasPlanes are defined. Needs to be validated somehow.

Reply:
This is still an open question that needs to be adressed

Reply:
I wrote a more detailed explanation below: it is correct to do it and it needs to happen in a camera frame.
The HillasPlanes have been always initialized in the horizon frame - this is the very last operation of the method, after the coordinate in the sky is obtained.
The planes related to this correction are instead the camera planes, which in parallel pointing are parallel to the TiltedFrame, but in divergent pointing not - therefore the reason for the correction.